apps/emqx/src/emqx_channel.erl

```diff
@@ -1523,29 +1586,22 @@ handle_info({unsubscribe, TopicFilters}, Channel) ->
             {ok, NChannel}
         end
     );
-handle_info({sock_closed, Reason}, Channel = #channel{conn_state = idle}) ->
-    shutdown(Reason, Channel);
-handle_info({sock_closed, Reason}, Channel = #channel{conn_state = connecting}) ->
-    shutdown(Reason, Channel);
-handle_info(
-    {sock_closed, Reason},
-    Channel =
-        #channel{
-            conn_state = ConnState,
-            clientinfo = ClientInfo,
-            conninfo = ConnInfo,
-            session = Session
-        }
-) when
+handle_info({sock_closed, normal}, Channel = #channel{conn_state = ConnState}) when
     ?IS_CONNECTED_OR_REAUTHENTICATING(ConnState)
 ->
-    {Intent, Session1} = session_disconnect(ClientInfo, ConnInfo, Session),
-    Channel1 = ensure_disconnected(Reason, maybe_publish_will_msg(sock_closed, Channel)),
-    Channel2 = Channel1#channel{session = Session1},
-    case maybe_shutdown(Reason, Intent, Channel2) of
-        {ok, Channel3} -> {ok, ?REPLY_EVENT(disconnected), Channel3};
-        Shutdown -> Shutdown
-    end;
+    %% normal disconnect(by client's DISCONNECT packet) and close socket
+    %% already traced `client.disconnect`, no need to trace `broker.disconnect`
+    process_maybe_shutdown(normal, Channel);
+handle_info({sock_closed, Reason}, Channel = #channel{conn_state = ConnState}) when
+    ?IS_CONNECTED_OR_REAUTHENTICATING(ConnState)
+->
+    %% Unexpected sock_closed when `connected` or `reauthenticating`
```

Comment:
nit: this is not unexpectd, `{sock_closed, Reason}` is initiated internally when sending data to a socket returned `{error, Reason}`.
```suggestion
    %% Socket closed when `connected` or `reauthenticating`
```